File: 백준/Silver/2579. 계단 오르기/계단 오르기.cc
#include <iostream>
#include <vector>

using namespace std;

int n;
int step[301];
int dp[301];

void input()
{
	cin >> n;
	for(int i=1; i<=n; i++)	cin >> step[i];
}

void solution()
{
	dp[1] = step[1];
	dp[2] = step[1] + step[2];
	dp[3] = max(step[1] + step[3], step[2] + step[3]);
	for(int i=4; i<=n; i++)
	{
		dp[i] = max(dp[i-3] + step[i-1] + step[i], dp[i-2] + step[i]);
	}
	cout << dp[n] << "\n";
}

void solve()
{
    input();
	solution();
}

int main()
{
    cin.tie(NULL);
    ios::sync_with_stdio(false);
//	freopen("input.txt", "r", stdin);
    solve();
    return 0;
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
# import sys
# sys.stdin = open("input.txt", "r")

n = int(input())
steps = [0 for _ in range(n+10)]
for i in range(1, n+1):
    steps[i] = int(input())
dp = [0 for _ in range(n+10)]

dp[1] = steps[1]
dp[2] = steps[1] + steps[2]
dp[3] = max(steps[1] + steps[3], steps[2] + steps[3])

for i in range(4, n+1):
    dp[i] = max(dp[i-2] + steps[i], dp[i-3] + steps[i-1] + steps[i])

print(dp[n])